extensions/amp-web-push/0.1/amp-web-push.service-worker.js

```diff
@@ -0,0 +1,155 @@
+};
+
+/*
+  According to
+  https://w3c.github.io/ServiceWorker/#run-service-worker-algorithm:
+
+  "user agents are encouraged to show a warning that the event listeners
+  must be added on the very first evaluation of the worker script."
+
+  We have to register our event handler statically (not within an
+  asynchronous method) so that the browser can optimize not waking up the
+  service worker for events that aren't known for sure to be listened for.
+
+  Also see: https://github.com/w3c/ServiceWorker/issues/1156
+*/
+self.addEventListener('message', event => {
+  /*
+    Messages sent from amp-web-push have the format:
+
+    - command: A string describing the message topic (e.g.
+      'amp-web-push-subscribe')
+
+    - payload: An optional JavaScript object containing extra data relevant to
+      the command.
+   */
+  const {command, payload} = event.data;
+
+  switch (command) {
+    case WorkerMessengerCommand.AMP_SUBSCRIPION_STATE:
+      onMessageReceivedSubscriptionState(payload);
+      break;
+    case WorkerMessengerCommand.AMP_SUBSCRIBE:
+      onMessageReceivedSubscribe(payload);
+      break;
+    case WorkerMessengerCommand.AMP_UNSUBSCRIBE:
+      onMessageReceivedUnsubscribe(payload);
+      break;
+  }
+});
+
+/**
+  Broadcasts a single boolean describing whether the user is subscribed.
+ */
+function onMessageReceivedSubscriptionState() {
+  let retrievedPushSubscription = null;
+  self.registration.pushManager.getSubscription()
+      .then(pushSubscription => {
+        retrievedPushSubscription = pushSubscription;
+        if (!pushSubscription) {
+          return null;
+        } else {
+          return self.registration.pushManager.permissionState(
+              pushSubscription.options
+        );
+        }
+      }).then(permissionStateOrNull => {
+        if (permissionStateOrNull == null) {
+          broadcastReply(WorkerMessengerCommand.AMP_SUBSCRIPION_STATE, false);
+        } else {
+          const isSubscribed = !!retrievedPushSubscription &&
+            permissionStateOrNull === 'granted';
+          broadcastReply(WorkerMessengerCommand.AMP_SUBSCRIPION_STATE,
+              isSubscribed);
+        }
+      });
+}
+
+/**
+  Subscribes the visitor to push.
+
+  The broadcast value is null (not used in the AMP page).
+ */
+function onMessageReceivedSubscribe() {
+  self.registration.pushManager.subscribe({
+    userVisibleOnly: true,
+    applicationServerKey: 'fake-demo-key',
+  }).then(() => {
+    // IMPLEMENT: Forward the push subscription to your server here
+    broadcastReply(WorkerMessengerCommand.AMP_SUBSCRIBE, null);
+  });
+}
+
+
+/**
+  Unsubscribes the subscriber from push.
+
+  The broadcast value is null (not used in the AMP page).
+ */
+function onMessageReceivedUnsubscribe() {
+  self.registration.pushManager.getSubscription()
+      .then(subscription => subscription.unsubscribe())
+      .then(() => {
+    // OPTIONALLY IMPLEMENT: Forward the unsubscription to your server here
```

Comment:
This is a marker for the reader reviewing a service worker implementation example. Users using push vendors shouldn't have to deal with this, as the vendor should be implementing this. For users with their own service worker, they'll most likely already have code subscribing and unsubscribing users, and they'll just take away the parts relevant to AMP (the messages accepted by the service worker and the kinds of replies to broadcast). The code for subscribing and unsubscribing users is more to show how and where the pieces fit together.